components/event-component-factory.js

```diff
@@ -0,0 +1,109 @@
+// component-factory.js
+
+var _ = require('underscore');
+var noflo = require('noflo');
+
+/**
+ * Creates a noflo Component factory function from a component definitation,
+ * registers any event handlers on definitation. Triggers ondata events for outPorts.
+ *
+ * Usage:
+ *  componentFactory({
+ *      description: "text",
+ *      icon: 'fontawesome-icon-name',
+ *      inPorts: {
+ *          portName: {
+ *              ondata: function(payload)
+ *          }
+ *      },
+ *      outPorts: {
+ *          portName: {
+ *              ondata: function(payload)
+ *          }
+ *      },
+ *      onicon: function(icon)
+ *   });
+ */
+module.exports = function(def){
+    if (!def) throw Error("No parameter");
+    return function() {
+        // noflo requires each port and nodeInstance to have its own options object
+        var nodeInstance = new noflo.Component({
+            outPorts: _.mapObject(def.outPorts, _.clone),
+            inPorts: _.mapObject(def.inPorts, _.clone)
+        });
+        triggerPortDataEvents(nodeInstance.outPorts);
+        registerPorts(nodeInstance.outPorts, def.outPorts);
+        registerPorts(nodeInstance.inPorts, def.inPorts);
+        registerListeners(nodeInstance, def);
+        nodeInstance.description = def.description;
+        nodeInstance.setIcon(def.icon);
+        return nodeInstance;
+    };
+};
+
+/**
+ * Fires data event when OutPort send function is called on port.
+ */
+function triggerPortDataEvents(ports) {
+    _.each(ports, function(port) {
+        port.send = _.partial(thenTrigger, port.send, 'data');
+    });
+}
+
+/**
+ * Emit event after function is called
+ * @param fn function to call
+ * @param event event to emit
+ */
+function thenTrigger(fn, event /* arguments to fn and event */) {
```

Comment:
This is to fire the 'data' event when the send function is called. This is to allow code (that will track output payloads) to register for data events on the output port.